crates/swc_ecma_minifier/src/compress/optimize/switches.rs

```diff
@@ -16,171 +14,146 @@ where
-            });
-            if should_preserve_switch {
-                // Prevent infinite loop.
-                if stmt.cases.len() == 1 {
-                    return;
+        let mut var_ids = vec![];
+        let mut cases = Vec::new();
+        let mut exact = None;
+
+        for (idx, case) in stmt.cases.iter_mut().enumerate() {
+            if let Some(test) = case.test.as_ref() {
+                if let Some(e) = is_primitive(tail_expr(test)) {
+                    if e.eq_ignore_span(tail) {
+                        cases.push(case.take());
+                        exact = Some(idx);
+                        break;
+                    } else {
+                        var_ids.extend(case.cons.extract_var_ids())
+                    }
+                } else {
+                    cases.push(case.take())
                 }
-
-                report_change!("switches: Removing unreachable cases from a constant switch");
             } else {
-                report_change!("switches: Removing a constant switch");
+                cases.push(case.take())
             }
+        }
 
-            self.changed = true;
-            let mut preserved = vec![];
-            if !should_preserve_switch && !discriminant.is_lit() {
-                preserved.push(Stmt::Expr(ExprStmt {
-                    span: stmt.span,
-                    expr: discriminant.take(),
-                }));
-
-                if let Some(expr) = stmt.cases[case_idx].test.take() {
-                    preserved.push(Stmt::Expr(ExprStmt {
-                        span: stmt.cases[case_idx].span,
-                        expr,
-                    }));
+        if let Some(exact) = exact {
+            let exact_case = cases.last_mut().unwrap();
+            let mut terminate = exact_case.cons.terminates();
+            for case in stmt.cases[(exact + 1)..].iter_mut() {
+                if terminate {
+                    var_ids.extend(case.cons.extract_var_ids())
+                } else {
+                    terminate |= case.cons.terminates();
+                    exact_case.cons.extend(case.cons.take())
                 }
             }
+            // remove default if there's an exact match
+            cases.retain(|case| case.test.is_some());
 
-            for case in &stmt.cases[..case_idx] {
-                for cons in &case.cons {
-                    var_ids.extend(
-                        cons.extract_var_ids()
-                            .into_iter()
-                            .map(|name| VarDeclarator {
-                                span: DUMMY_SP,
-                                name: Pat::Ident(name.into()),
-                                init: None,
-                                definite: Default::default(),
-                            }),
-                    );
+            if cases.len() == 2 {
+                let last = cases.last_mut().unwrap();
+
+                self.changed = true;
```

Comment:
Can you add a `report_change!` here?

Reply:
This would be reported in following pass which change two cases switch into if else